### PR TITLE
core - refactor action handling

### PR DESCRIPTION
**Pull request by tjstansell**
This should address points in issue #3531:
  - a resource should be guaranteed to only move to followup actions upon success
  - some actions should allow accessing failed resources (e.g. notify)
  - uncaught exceptions in actions should not affect resources that
    previously succeeded.  they should continue to followup actions.

This PR is not complete yet... but a good starting point for review.  For the resources I've adjusted, tests are passing (except for those that look for exceptions that no longer happen).

### CODE REVIEW

File: c7n/actions/core.py [hunk 1]
```diff
@@ -55,6 +57,71 @@ def factory(self, data, manager):
...
+    def ok(self, resources):
+        self._add(resources, status="ok")
+
+    def skip(self, resources, reason):
+        self._add(resources, status="skip", reason=reason)
+
+    def error(self, resources, reason):
+        self._add(resources, status="error", reason=reason)
+
+    def exception(self, e):
```
Comment (kapilt) [inline c7n/actions/core.py:87]: whats the delta between error and exception? .. exception classes don't per se clearly serialize

**Comment by kapilt** [comment]
following up from gitter conversation. Generally for larger prs, especially those refactoring around a lot of existing semantics, it would be good to have a dedicated proposal issue that works out the interface spelling and changes (see examples https://github.com/cloud-custodian/cloud-custodian/issues?q=is%3Aissue+is%3Aopen+label%3Aarea%2Fproposal)  that allows for discussion and iteration, and then a smaller pr that tries to focus on the core semantic, before applying it over a series of prs. a large pr with a lot of unrelated changes is generally going to be much harder to review and carries much more risk and is going to grow stale.  re extraneous changes, i see unrelated signature and concurrency changes in multiple places in this pr. wrt to the general approach, it seems pretty reasonable, but it would be good to discuss on a proposal issue.

**Comment by tjstansell** [comment]
yeah.  i got carried away after making the initial changes i wanted by actually trying to see how they would be used ... and applying some consistency across how all of the actions are actually run.  as for the proposal, that's what the discussion in #3531 was.  I submitted a proposal a week ago and didn't see any responses, so I went to start implementation and see what it would really look like once I got out of the clouds.  I'm going to update this PR today with just the core changes.  It should be compatible enough in that it shouldn't really change existing behavior.  Then I can submit other PRs to update all of the resources to use the new framework, probably in batches rather than 1 PR per resource type.

File: c7n/utils.py [hunk 2]
```diff
@@ -169,7 +169,7 @@ def default(self, obj):
         try:
             return json.JSONEncoder.default(self, obj)
         except TypeError:
-            if getattr(obj, '__dict__'):
+            if hasattr(obj, '__dict__'):
```
Comment (kapilt) [inline c7n/utils.py:172]: hasattr swallows exceptions, getattr with a default won't

Comment (tjstansell) [reply c7n/utils.py:172]: will that somehow affect the current exception that i'd be re-raising if __dict__ is not an attribute, then?  so it'd be better to use `if getattr(obj, '__dict__', None) is not None:`?

**Comment by tjstansell** [comment]
in theory i'm done making changes to this PR... it shouldn't have any extraneous changes now too... so I'd love to get a review ...

**Comment by tjstansell** [comment]
This is highly stale and doesn't sound like we have agreement on the methodology here, so I'm going to just close this PR.  If/when we revisit, we can start a new PR.

**Pull request closed without merging**